src/network/cares/cares.cpp

```diff
@@ -0,0 +1,207 @@
+      // atomic change failed, thus it was already initialized
+      log_->debug("C-ares library got initialized more than once");
+    } else if (auto status = ::ares_library_init(ARES_LIB_INIT_ALL);
+               ARES_SUCCESS != status) {
+      log_->error("Unable to initialize c-ares library - {}",
+                  ::ares_strerror(status));
+      // on library initialization failure we set initialized = false
+      assert(initialized_.compare_exchange_strong(expected, false));
+    }
+  }
+
+  Ares::~Ares() {
+    initialized_.store(false);
+    ares_library_cleanup();
+  }
+
+  void Ares::resolveTxt(
+      const std::string &uri,
+      const std::weak_ptr<boost::asio::io_context> &io_context,
+      Ares::TxtCallback callback) {
+    if (not initialized_.load()) {
+      log_->debug(
+          "Unable to execute DNS TXT request to %s due to c-ares library is "
+          "not initialized",
+          uri);
+      reportError(io_context, std::move(callback), Error::NOT_INITIALIZED);
+      return;
+    }
+    int status{ARES_SUCCESS};
+    ::ares_options options{};
+    ::ares_channel channel{nullptr};
+    options.timeout = 30'000;
+    status = ares_init_options(&channel, &options, ARES_OPT_TIMEOUTMS);
+    if (ARES_SUCCESS != status) {
+      log_->debug("Unable to initialize c-ares channel for request to {} - {}",
+                  uri, ::ares_strerror(status));
+      reportError(io_context, std::move(callback), Error::CHANNEL_INIT_FAILURE);
+      return;
+    }
+
+    auto request =
+        std::make_shared<RequestContext>(io_context, uri, std::move(callback));
+    requests_.push_back(request);
+
+    try {
+      std::thread worker([channel, request] {
+        ::ares_query(channel, request->uri.c_str(), ns_c_in, ns_t_txt,
+                     Ares::txtCallback, request.get());
+        Ares::waitAresChannel(channel);
+        ::ares_destroy(channel);
+      });
+      worker.detach();
+    } catch (std::runtime_error &e) {
+      log_->error("Ares unable to start worker thread - {}", e.what());
+      request->callback(Error::THREAD_FAILED);
+    }
+  }
+
+  void Ares::reportError(
+      const std::weak_ptr<boost::asio::io_context> &io_context,
+      Ares::TxtCallback callback, Ares::Error error) {
+    if (auto ctx = io_context.lock()) {
+      ctx->post([callback{std::move(callback)}, error] { callback(error); });
+      return;
+    }
+    log_->debug("IO context has expired");
+  }
+
+  void Ares::txtCallback(void *arg, int status, int, unsigned char *abuf,
+                         int alen) {
+    auto *request_ptr{static_cast<RequestContext *>(arg)};
+    auto processError = [request_ptr](int status_code) -> bool {
+      if (ARES_SUCCESS != status_code) {
+        reportError(request_ptr->io_context, request_ptr->callback,
+                    kQueryErrors.at(status_code));
+        removeRequest(request_ptr);
+        return true;
+      }
+      return false;
+    };
+    if (processError(status)) {
```

Comment:
It's not clear from the code how it works. Suggest to check status explicitly before error processing.